### PR TITLE
Togglable Mobile support

### PR DESCRIPTION
**Pull request by morenol**
**TL;DR -** [ Display openassessment units in mobile apps. ]

This is the same of https://github.com/edx/edx-ora2/pull/1134, but rebased on top of master and now, the feature is togglable through a  feature flag.

**Currently when we try to display an ora2 unit in mobile:**

image

**Once this change gets installed in edx-platform**
out

## Context

This is the piece of code of edx-platform that checks if an xblock supports mobile apps: https://github.com/edx/edx-platform/blob/61e1eda20df2825a409db3e2d36c69d7c36d3e2d/lms/djangoapps/course_api/blocks/transformers/student_view.py#L54

**What changed?**

- openassessment unit now will be displayed in the mobile apps if any of the feature flags for that are enabled (openresponseassessment.mobile_support waffle switch or ENABLE_ORA_MOBILE_SUPPORT edx-platform feature )

**How to test?**

* Install this xblock in edx-platform
* Configure the platform properly to support mobile apps
* turn on openresponseassessment.mobile_support waffle switch
* Build a mobile app that uses that sandboxed environment
* Test that a unit with an openassessment unit gets rendered in the app.

**Developer Checklist**
- [ ] Reviewed the [release process](./release_process.md)
- [x] ~Translations up to date~ (NA)
- [x] ~JS minified, SASS compiled~ (NA)
- [ ] Bumped version number in [setup.py](../setup.py) and [package.json](../package.json)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @morenol! I've created [OSPR-4802](https://openedx.atlassian.net/browse/OSPR-4802) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

**Comment by natabene** [comment]
@morenol Thank you for your contribution. Please let me know once it is ready for our review.

**Comment by morenol** [comment]
Hi @natabene,  this can be reviewed now

**Comment by morenol** [comment]
@pomegranited thank you for your review.

Regarding the first problem I only tested in Android and I was able to submit answers, I am going to try to test with iOS to see if I can replicate that error.


> On another topic, do you have any idea why the ORA2 xblock that's installed with the Demo course wouldn't show up in the mobile app? This course was imported before I ran your ORA2 XBlock version. I've tried re-publishing, but it still won't show up.

In Android that xblock is displayed in the mobile app but only after I removed the cache from the mobile app, Could you try to remove the cache of the app or something like that to see if that is the problem? Again, I am going to try to test that on iOS.

File: openassessment/xblock/config_mixin.py [hunk 1]
```diff
@@ -80,11 +81,14 @@ def is_feature_enabled(self, flag):
...
+        try:
+            if hasattr(self, 'location') and self._course_waffle_flag(flag).is_enabled(self.location.course_key):
+                return True
 
-        if self._waffle_switch(flag).is_enabled():
-            return True
+            if self._waffle_switch(flag).is_enabled():
+                return True
+        except ImportError:
+            pass
```
[the remark this answers is not in the document]
Comment (morenol) [reply openassessment/xblock/config_mixin.py:91]: I tried that, but since the decorator is not wrapping anything, everything from the new decorator is run when OpenAssessmentBlock is imported. So, it is raising import error in the tests that
 has this line `from openassessment.xblock.openassessmentblock import OpenAssessmentBlock`. I could move these imports to the places where OpenAssessmentBlock is used in each test and mock that import_waffle_switch, but I was not sure if that was better than just this change. Let me know if you think that it would be better at that way.


On the other hand, I think that I cannot wrap in the decorator the part that uses the waffle switches because [the part](https://github.com/edx/edx-platform/blob/61e1eda20df2825a409db3e2d36c69d7c36d3e2d/lms/djangoapps/course_api/blocks/transformers/student_view.py#L54) in the platform that checks if a Xblock has mobile support needs that  xblocks.support('multi_device') is ran on import time.

[the remark this answers is not in the document]
Comment (pomegranited) [reply openassessment/xblock/config_mixin.py:91]: @morenol Ah thank you for explaining, I see now why it's complicated.

Could you move this `try`/`catch` block closer to the source of the issue, and add comments explaining? That way, it's clear that the exception is directly related to this feature:

```diff
--- a/openassessment/xblock/mobile.py
+++ b/openassessment/xblock/mobile.py
@@ -15,8 +15,12 @@ class TogglableMobileSupport(ConfigMixin):
         """
         Add mobile support if the feature flag is enabled.
         """
-        if self.is_mobile_support_waffle_enabled:
-            return XBlock.supports('multi_device')(view)
+        try:
+            if self.is_mobile_support_waffle_enabled:
+                return XBlock.supports('multi_device')(view)
+        except ImportError:
+            # openedx environment not available, e.g. in testing
+            pass
         return view
```

**Comment by pomegranited** [comment]
@morenol 

> In Android that xblock is displayed in the mobile app but only after I removed the cache from the mobile app and restarted the lms and memcached containers.

I had to do those 3 steps too to make it work.. didn't know the app would cache so aggressively!  Thanks for the explanation.

**Comment by pomegranited** [comment]
@morenol Let me know when you're ready for me to re-test on iOS?

**Comment by morenol** [comment]
> @morenol Let me know when you're ready for me to re-test on iOS?

Thanks, @pomegranited. I will let you know

**Comment by natabene** [comment]
@pomegranited Do you want to be a cc-reviewer for this PR?

**Comment by pomegranited** [comment]
@natabene I'd love to, but I'm not a CC for edx-ora2 😄

**Comment by pomegranited** [comment]
@natabene  Oh! Looks like I am a CC for edx-ora2 after all!  Yes please!

**Comment by andrey-canon** [comment]
@morenol @pomegranited The submit error doesn't belong to ora, the ios app doesn't show alert dialogs, hence I opened this pr https://github.com/edx/edx-app-ios/pull/1401 you can test your ora changes with this version of the app.

**Comment by morenol** [comment]
@pomegranited I would like to add that we were not able to replicate your error, in our case we were able to save the responses but we had errors when trying to do the submission.

ios-error

**Comment by pomegranited** [comment]
Thanks @andrey-canon and @morenol ! Running with https://github.com/edx/edx-app-ios/pull/1401 got this working for me 👍 

- [x] I tested this on my devstack and the iOS app, running https://github.com/edx/edx-app-ios/pull/1401.
   - Tested that without enabling the `openresponseassessment.mobile_support` waffle switch: ensured that default, mobile-disabled behaviour is maintained.
   - Enabled the `openresponseassessment.mobile_support` waffle switch, and [restarted memcached, ran `generate_course_blocks`, and logged out and back into the iOS app](https://github.com/edx/edx-ora2/pull/1445#issuecomment-671630071): ensured that the ORA2 component displays in the iOS app, and that text and file upload submissions can be made.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A, mobile apps
- [x] Includes documentation for the feature flag.

**Comment by pomegranited** [comment]
@morenol Before we can merge, you'll need to merge the latest `master` and update the [`version`](https://github.com/eduNEXT/edx-ora2/blob/lmm/mobile/setup.py#L39), so the release can be tagged.

@nedbat This is my first core-committer review for ORA2, so could you confirm that once ^ is done, it's ok for me to merge this? CC @feanil @natabene

**Review by pomegranited — APPROVED** [review]
Perfect, thank you @morenol !

Approved, see [testing steps above](https://github.com/edx/edx-ora2/pull/1445#issuecomment-703217338).

**Comment by pomegranited** [comment]
Sorry @morenol , I didn't merge this in time and now there's a conflict on setup.py (probably just the version change). Could you fix that and I'll merge it ASAP?

**Comment by edx-webhook** [comment]
@morenol 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-ora2&entry.867055334=https://github.com/edx/edx-ora2/pull/1445&entry.1484655318=https://github.com/morenol&entry.752974735=2020-07-06+19:01&entry.1917517419=2020-10-14+23:46&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by pomegranited** [comment]
Awesome, thank you for building this support @morenol !
